src/libraries/Microsoft.Extensions.Options/src/OptionsSnapshot.cs

```diff
@@ -0,0 +1,66 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+
+namespace Microsoft.Extensions.Options
+{
+    /// <summary>
+    /// Implementation of <see cref="IOptionsSnapshot{TOptions}"/>.
+    /// </summary>
+    /// <typeparam name="TOptions">Options type.</typeparam>
+    internal sealed class OptionsSnapshot<[DynamicallyAccessedMembers(Options.DynamicallyAccessedMembers)] TOptions> :
+        IOptionsSnapshot<TOptions>
+        where TOptions : class
+    {
+        private readonly IOptionsMonitor<TOptions> _optionsMonitor;
+
+        private volatile ConcurrentDictionary<string, TOptions> _cache;
+        private volatile TOptions _unnamedOptionsValue;
+
+        /// <summary>
+        /// Initializes a new instance with the specified options configurations.
+        /// </summary>
+        /// <param name="optionsMonitor">The options monitor to use to provide options.</param>
+        public OptionsSnapshot(IOptionsMonitor<TOptions> optionsMonitor)
+        {
+            _optionsMonitor = optionsMonitor;
+        }
+
+        /// <summary>
+        /// The default configured <typeparamref name="TOptions"/> instance, equivalent to Get(Options.DefaultName).
+        /// </summary>
+        public TOptions Value => Get(Options.DefaultName);
+
+        /// <summary>
+        /// Returns a configured <typeparamref name="TOptions"/> instance with the given <paramref name="name"/>.
+        /// </summary>
+        public TOptions Get(string name)
+        {
+            if (name == null || name == Options.DefaultName)
+            {
+                if (_unnamedOptionsValue is TOptions value)
+                {
+                    return value;
+                }
+
+                return _unnamedOptionsValue = _optionsMonitor.Get(Options.DefaultName);
+            }
+
+            var cache = _cache ?? Interlocked.CompareExchange(ref _cache, new(concurrencyLevel: 1, capacity: 5, StringComparer.Ordinal), null) ?? _cache;
```

Comment:
@stephentoub can you look here? I think it's copied from the other pattern we use in Options\<T\>

Reply:
The pattern looks fine.  It's saying:
- Volatile read _cache: if it's not null, use it.
- It's null, so create a new one and atomically set _cache to a new instance iff _cache is still null.
- If it wasn't still null, use what it was instead.
- If it was still null, it's now instead the new instance we set it to, so use that.

Reply:
I'm waiting for your book to come out. Concurrency patterns with @stephentoub